### PR TITLE
add all

### PR DESCRIPTION
**Pull request by na206851**
review

### CODE REVIEW

File: src/main/zlatopolsky/chapter1/Ch1N1_3.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package zlatopolsky.chapter1;
+
+import java.util.Scanner;
+
+public class Ch1N1_3 {
+    public void printNumber() {
+        Scanner sc = new Scanner(System.in);
+        int number = sc.nextInt();
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter1/Ch1N1_3.java]: //comment

**Comment by oshavrov** [comment]
All the code written in Java should be in "java" folder, like
src/main/java/zlatopolskiy/ArrayUtils.java
instead of
src/main/zlatopolskiy/ArrayUtils.java
Same for tests

File: src/main/zlatopolsky/chapter10/Ch1N10_53.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package zlatopolsky.chapter10;
+
+import java.util.Scanner;
+
+public class Ch1N10_53 {
+
+    public void reverseRow(Scanner sc) {
```
Comment (oshavrov) [inline src/main/zlatopolsky/chapter10/Ch1N10_53.java]: What this method does?
I tried to launch it, but it is just constantly asking for an input, and does nothing

File: src/main/zlatopolsky/chapter10/Ch1N10_51.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package zlatopolsky.chapter10;
+
+public class Ch1N10_51 {
+
+    public int recursionExpressionA(int n) {
```
Comment (oshavrov) [inline src/main/zlatopolsky/chapter10/Ch1N10_51.java]: It should print the results like:
5
4
3
2
1

File: src/main/zlatopolsky/chapter10/Ch1N10_54.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    public ArrayList translation10to2(int n, ArrayList list10to2) {
+        if (n == 0) {
+            list10to2.add(0);
+            return list10to2;
+        } else if (n == 1) {
+            list10to2.add(1);
+            return list10to2;
+        } else {
+            list10to2.add(n % 2);
+            return (translation10to2(n / 2, list10to2));
```
Comment (oshavrov) [inline src/main/zlatopolsky/chapter10/Ch1N10_54.java]: Redundant braces

File: src/main/zlatopolsky/chapter10/Ch1N10_53.java (hunk 2 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_53.java]: fix!: corrected name method, this method recursion reverse row while will not be zero

File: src/main/zlatopolsky/chapter10/Ch1N10_51.java (hunk 3 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_51.java]: yes it error, i'm fix the method

File: src/main/zlatopolsky/chapter10/Ch1N10_54.java (hunk 4 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_54.java]: fix in all files

File: src/main/zlatopolsky/chapter10/Ch1N10_43.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+    public static int sumDigits(int x) {
+        if (x <= 0) {
+            return x;
+        }
+        return x % 10 + (sumDigits(x / 10));
+    }
+
+    public static int sumDigits(int x, int count) {
+        if (x <= 0) {
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter10/Ch1N10_43.java]: //code duplication, int count - you didn't use this variable.

File: src/main/zlatopolsky/chapter10/Ch1N10_44.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        }
+        return digitalRoot(sumDigits(x));
+    }
+
+    public int sumDigits(int x) {
+        if (x < 10) {
+            return x;
+        }
+        return x % 10 + sumDigits(x / 10);
+    }
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter10/Ch1N10_44.java]: //You can combine these methods and create digitalRoot method

File: src/main/zlatopolsky/chapter10/Ch1N10_52.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package zlatopolsky.chapter10;
+
+public class Ch1N10_52 {
+
+    public int reverseNumber(int n) {
+        if (n < 10) {
+            return n;
+        } else {
+            return Integer.parseInt(String.valueOf(n % 10) + reverseNumber(n / 10));
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter10/Ch1N10_52.java]: //Why you use Integer.parseInt and String.valueOf? You have got an integer n,  you can use just n.

File: src/main/zlatopolsky/chapter11/Ch1N11_158.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
+package zlatopolsky.chapter11;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class Ch1N11_158 {
+
+    public int[] removeDuplicate(int[] nums) {
+        nums = Arrays.stream(nums).distinct().toArray();
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter11/Ch1N11_158.java]: // I think you need to learn about HashSet

File: src/main/zlatopolsky/chapter4/Ch1N4_106.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package zlatopolsky.chapter4;
+
+public class Ch1N4_106 {
+
+    public String season(int month) {
+        String season = "";
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter4/Ch1N4_106.java]: // I think you can remove variable String season and will do return "winter" etc.

File: src/main/zlatopolsky/chapter4/Ch1N4_15.java [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package zlatopolsky.chapter4;
+
+public class Ch1N4_15 {
+
+    public int humanAge(int yearOfBirth, int monthOfBirth, int yearCurrent, int monthCurrent) {
+        return ((yearCurrent * 12 +
+                monthCurrent) - (yearOfBirth * 12 + monthOfBirth)) / 12;
```
Comment (EmmaBuraeva) [inline src/main/zlatopolsky/chapter4/Ch1N4_15.java]: // You need delete 12, because you multiply and divide  by 12

File: src/main/zlatopolsky/chapter10/Ch1N10_51.java (hunk 3 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_51.java]: fix

File: src/main/zlatopolsky/chapter10/Ch1N10_43.java (hunk 5 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_43.java]: yes, fix error and rename method

File: src/main/zlatopolsky/chapter10/Ch1N10_52.java (hunk 7 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_52.java]: this is necessary so that the numbers do not add up

File: src/main/zlatopolsky/chapter11/Ch1N11_158.java (hunk 8 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter11/Ch1N11_158.java]: yes, i agree

File: src/main/zlatopolsky/chapter4/Ch1N4_15.java (hunk 10 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter4/Ch1N4_15.java]: test AgeHuman
no, i test your option, it produced uncorrected information

File: src/main/zlatopolsky/chapter4/Ch1N4_106.java (hunk 9 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter4/Ch1N4_106.java]: i'm delete case 1, 2, 12

File: src/main/zlatopolsky/chapter10/Ch1N10_52.java (hunk 7 above)
Comment (oshavrov) [reply src/main/zlatopolsky/chapter10/Ch1N10_52.java]: Ho-ho, you can multiply the number by 10 before adding the new digit

Comment (oshavrov) [reply src/main/zlatopolsky/chapter10/Ch1N10_52.java]: Solving this task with numbers is preferable. Conversion to String and back to Integer is a cheat

File: src/main/zlatopolsky/chapter11/Ch1N11_158.java (hunk 8 above)
Comment (oshavrov) [reply src/main/zlatopolsky/chapter11/Ch1N11_158.java]: There is no need to assign result to "nums". Just return the value

Comment (oshavrov) [reply src/main/zlatopolsky/chapter11/Ch1N11_158.java]: And also try to avoid reassigning to input parameters

File: src/main/java/zlatopolsky/chapter10/Ch1N10_54.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    public int reverseArrayList(ArrayList list10to2) {
+        int i = list10to2.size();
+        StringBuilder number = new StringBuilder();
+        while (i > 0) {
+            number.append(list10to2.get(i - 1));
+            i--;
+        }
+        return Integer.parseInt(number.toString());
+    }
+    public ArrayList translation10to2(int n, ArrayList list10to2) {
```
Comment (oshavrov) [inline src/main/java/zlatopolsky/chapter10/Ch1N10_54.java]: It is better to have a blank line between 2 methods

File: src/main/java/zlatopolsky/chapter10/Ch1N10_55.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        int i = reverseList.size();
+        StringBuilder number = new StringBuilder();
+        while (i > 0) {
+            number.append(reverseList.get(i - 1));
+            i--;
+        }
+        return number.toString();
+    }
+
+    public ArrayList tenToTwo(int number, ArrayList list10to2) {
```
Comment (oshavrov) [inline src/main/java/zlatopolsky/chapter10/Ch1N10_55.java]: Let's declare method signatures at interface level. Use `List<Integer>` instead of ArrayList. 
Mind the parameter of the type

File: src/main/java/zlatopolsky/chapter11/Ch1N11_158.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Arrays;
+
+public class Ch1N11_158 {
+
+    public int[] removeDuplicate(int[] nums) {
+        nums = Arrays.stream(nums).distinct().toArray();
+        return nums;
+    }
+
+    public ArrayList<Integer> removeDuplicate2(int[] nums) {
```
Comment (oshavrov) [inline src/main/java/zlatopolsky/chapter11/Ch1N11_158.java]: Good catch with the type parameter. Please also move to an interface level: `List<Integer>`

File: src/main/java/zlatopolsky/chapter12/Ch1N12_63.java [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    public int[][] numberOfStudents(int[][] arr) {
+        Random children = new Random();
+        for (int i = 0; i < arr.length; i++) {
+            for (int j = 0; j < arr[0].length; j++) {
+                arr[i][j] = children.nextInt(30 - 15) + 15;
+            }
+        }
+        return arr;
+    }
+    public int[] columnAverage(int[][] arr) {
```
Comment (oshavrov) [inline src/main/java/zlatopolsky/chapter12/Ch1N12_63.java]:
```
}
    public
```
Search in the project by this template, and put a line between methods

File: src/main/java/zlatopolsky/chapter13Employee/Database.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        ArrayList<Employee> result = new ArrayList<>();
+        for (Employee employee : employees) {
+            if (yearInCompany == employee.getYearInCompany()) {
+                result.add(employee);
+            }
+        }
+        return result;
+    }
+
+    public ArrayList<Employee> searchEmployee(String substring) {
```
Comment (oshavrov) [inline src/main/java/zlatopolsky/chapter13Employee/Database.java]: Again, `List<Employee>`, later you'll get to know, why am I asking to use interfaces

File: src/test/java/zlatopolsky/chapter13Employee/DatabaseTest.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
+package java.zlatopolsky.chapter13Employee;
+
+class DatabaseTest {
+    public static void main(String[] args) {
+        Database database = new Database();
+        database.addEmployee(new Employee("bob", "lastName", "surname", 2020, 4));
+        database.addEmployee(new Employee("nikita", "secondLasbotName", "seson", 2014, 9));
+        database.addEmployee(new Employee("nikita", "bobob", 2021, 9));
+        // System.out.println(database.getEmployeesByYearsInCompany(2));
```
Comment (oshavrov) [inline src/test/java/zlatopolsky/chapter13Employee/DatabaseTest.java]: Let's avoid stale comments. 
This method has to be valid. Use it to check values

File: src/test/java/zlatopolsky/chapter3/Ch1N3_26Test.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+        Ch1N3_26 test=new Ch1N3_26();
+        System.out.println(test.expressionA(true, true, true));
+        System.out.println(test.expressionA(true, true, false));
+        System.out.println(test.expressionA(true, false, false));
+        System.out.println(test.expressionA(false, false, false));
+        System.out.println(test.expressionA(false, true, true));
+        System.out.println(test.expressionA(false, false, true));
+        System.out.println(test.expressionA(false, true, false));
+
+
```
Comment (oshavrov) [inline src/test/java/zlatopolsky/chapter3/Ch1N3_26Test.java]: One line is enough

File: src/main/zlatopolsky/chapter10/Ch1N10_52.java (hunk 7 above)
Comment (na206851) [reply src/main/zlatopolsky/chapter10/Ch1N10_52.java]: fix method reverse, use only mathematical operation

File: src/test/java/zlatopolsky/chapter3/Ch1N3_26Test.java (hunk 17 above)
Comment (na206851) [reply src/test/java/zlatopolsky/chapter3/Ch1N3_26Test.java]: fix

File: src/main/java/zlatopolsky/chapter10/Ch1N10_54.java (hunk 11 above)
Comment (na206851) [reply src/main/java/zlatopolsky/chapter10/Ch1N10_54.java]: yes , add empty line

File: src/java/main/zlatopolsky/chapter10/Ch1N10_54.java [hunk 18]
```diff
@@ -0,0 +1,29 @@
+package zlatopolsky.chapter10;
+
+import java.util.List;
+
+public class Ch1N10_54 {
+
+    public int reverseArrayList(List<Integer> list10to2) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter10/Ch1N10_54.java]: It is just a List. Let's be more general in method names

File: src/java/main/zlatopolsky/chapter10/Ch1N10_54.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+        int i = list10to2.size();
+        StringBuilder number = new StringBuilder();
+        while (i > 0) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter10/Ch1N10_54.java]: You are still using Strings with numeric operations. 
Don't use any String-related structures here, only numbers

File: src/java/main/zlatopolsky/chapter11/Ch1N11_158.java [hunk 20]
```diff
@@ -0,0 +1,50 @@
+package zlatopolsky.chapter11;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class Ch1N11_158 {
+
+    public int[] removeDuplicate(int[] nums2) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter11/Ch1N11_158.java]: Probably, it should be in plural form: "removeDuplicates", because we handle the input with more than one duplicate

File: src/java/main/zlatopolsky/chapter11/Ch1N11_158.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        return Arrays.stream(nums2).distinct().toArray();
+    }
+
+    public List<Integer> removeDuplicate2(int[] nums) {
+        List<Integer> list = new ArrayList<>();
+        int index = 0;
+        int count = 0;
+        int a = 0;
+        while (a < nums.length) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter11/Ch1N11_158.java]: for loop is more suitable here

File: src/java/main/zlatopolsky/chapter11/Ch1N11_158.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+                    nums[index] = nums[i];
+                } else {
+                    nums[index] = 0;
+                }
+            }
+            index++;
+        }
+        index = 0;
+
+        while (index < nums.length) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter11/Ch1N11_158.java]: Same here, you are sharing the variable "index" between two operators. It is error prone

Comment (oshavrov) [reply src/java/main/zlatopolsky/chapter11/Ch1N11_158.java]: Just use 2 different for loops

File: src/java/main/zlatopolsky/chapter11/Ch1N11_245.java [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    public int[] sortAndCopyArray(int nums[]) {
+        int[] result = new int[nums.length];
+        int index = 0;
+        int lastIndex = result.length - 1;
+
+        for (int i = 0; i < result.length; i++) {
+            if (nums[i] >= 0) {
+                result[lastIndex] = nums[i];
+                lastIndex--;
+            } else if (nums[i] <= 0) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter11/Ch1N11_245.java]: The upper ">=" already covers numbers greater or equal 0. You can put "<", this will be enough

File: src/java/main/zlatopolsky/chapter12/Ch1N12_28.java [hunk 24]
```diff
@@ -0,0 +1,46 @@
+package zlatopolsky.chapter12;
+
+public class Ch1N12_28 {
+    /**
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter12/Ch1N12_28.java]: Remove the comment, it is not related to the problem

File: src/java/main/zlatopolsky/chapter12/Ch1N12_28.java [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+                matrix[i][colon] = count++;
+            }
+            colon += 1;
+
+        }
+        print(matrix);
+        return matrix;
+    }
+
+    public void print(int[][] matrix) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter12/Ch1N12_28.java]: Don't you have a common method for all the print operations on matrices?

File: src/java/main/zlatopolsky/chapter13Employee/Employee.java [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+
+    public int getStartJobYear() {
+        return startJobYear;
+    }
+
+    public int getStartJobMonth() {
+        return startJobMonth;
+    }
+
+    Employee(String firstName, String lastName, int startJobYear, int startJobMonth) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter13Employee/Employee.java]: Let's create these constructors also "public"

File: src/java/main/zlatopolsky/chapter2/Ch1N2_39.java [hunk 27]
```diff
@@ -0,0 +1,11 @@
+package zlatopolsky.chapter2;
+
+public class Ch1N2_39 {
+    public int TimeToDegrees(int hour, int min) {
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter2/Ch1N2_39.java]: Methods in Java are named starting with lower case letter usually

File: src/java/main/zlatopolsky/chapter9/Ch1N9_185.java [hunk 28]
```diff
@@ -0,0 +1,19 @@
+package zlatopolsky.chapter9;
+
+public class Ch1N9_185 {
+
+    public int validPairs(String str) {
+        char[] arr = str.toCharArray();
+        int count = 0;
+        char prev = ' ';
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter9/Ch1N9_185.java]: How do you use the "prev" variable?
Please check how this method works on these examples:
Valid: "(())", "()()", "()"
Invalid: ")(", "(()", "()(()"

File: src/java/main/zlatopolsky/chapter9/Ch1N9_42.java [hunk 29]
```diff
@@ -0,0 +1,9 @@
+package zlatopolsky.chapter9;
+
+public class Ch1N9_42 {
+
+    public String reverseString(String str) {
+        StringBuilder strReverse = new StringBuilder(str);
+        return strReverse.reverse().toString().toLowerCase();
```
Comment (oshavrov) [inline src/java/main/zlatopolsky/chapter9/Ch1N9_42.java]: Why do you lower the string? Is it a part of requirements?

File: src/java/test/zlatopolsky/chapter9/Ch1N9_185Test.java [hunk 30]
```diff
@@ -0,0 +1,9 @@
+package zlatopolsky.chapter9;
+
+class Ch1N9_185Test {
+    public static void main(String[] args) {
+        Ch1N9_185 test = new Ch1N9_185();
+        String str = ")(";
+        System.out.println(test.validPairs(str));
```
Comment (oshavrov) [inline src/java/test/zlatopolsky/chapter9/Ch1N9_185Test.java]: More test cases, positive, negative

**Review by oshavrov — CHANGES_REQUESTED** [review]

File: src/java/main/zlatopolsky/chapter10/Ch1N10_54.java (hunk 19 above)
Comment (na206851) [reply src/java/main/zlatopolsky/chapter10/Ch1N10_54.java]: fix! delete StringBuilder, create new method translate10to2

File: src/java/main/zlatopolsky/chapter2/Ch1N2_39.java (hunk 27 above)
Comment (na206851) [reply src/java/main/zlatopolsky/chapter2/Ch1N2_39.java]: fix rename method

File: src/java/main/zlatopolsky/chapter12/Ch1N12_28.java (hunk 25 above)
Comment (na206851) [reply src/java/main/zlatopolsky/chapter12/Ch1N12_28.java]: no, for each case i write new operator , for independence

File: src/java/main/zlatopolsky/chapter9/Ch1N9_42.java (hunk 29 above)
Comment (na206851) [reply src/java/main/zlatopolsky/chapter9/Ch1N9_42.java]: no , this was not in expression, i think this is more correct

File: src/java/main/zlatopolsky/chapter9/Ch1N9_185.java (hunk 28 above)
Comment (na206851) [reply src/java/main/zlatopolsky/chapter9/Ch1N9_185.java]: this error, i write new solution and tested edge cases

File: src/java/test/zlatopolsky/chapter9/Ch1N9_185Test.java (hunk 30 above)
Comment (na206851) [reply src/java/test/zlatopolsky/chapter9/Ch1N9_185Test.java]: yes, i'm tested method with edge cases
